utils/queuer.go

```diff
@@ -67,13 +96,19 @@ func (q GeneralKVQueue) Dequeue(value codec.ProtoMarshaler, filter ...func(value
 		return false
 	}
 
-	if len(filter) > 0 && !filter[0](value) {
-		return false
-	}
+	isQualified, continueIfNotQualified := filter(value)
+	switch {
+	case isQualified:
+		q.store.Delete(KeyFromBz(iter.Key()))
 
-	q.store.Delete(KeyFromBz(iter.Key()))
+		return true
+	case continueIfNotQualified:
+		iter.Next()
 
-	return true
+		return q.dequeue(value, iter, filter)
```

Comment:
Don't think golang does tail call optimization, in which case, use a loop instead of recursion?